sklearn/utils/estimator_checks.py

```diff
@@ -1915,6 +1917,42 @@ def transform(self, X):
                     shallow_params.items()))
 
 
+@ignore_warnings(category=(DeprecationWarning, FutureWarning))
+def check_set_params(name, estimator_orig):
+    # Check that get_params() returns the same thing
+    # before and after set_params() with some fuzz
+    estimator = clone(estimator_orig)
+
+    params = estimator.get_params()
+    msg = ("get_params result does not match what was passed to set_params")
+
+    estimator.set_params(**params)
+    new_params = estimator.get_params()
+    assert_equal(params.keys(), new_params.keys(), msg)
+    for k, v in new_params.items():
+        assert_is(params[k], v, msg)
+
+    # some fuzz values
+    test_values = [-np.inf, np.inf, None]
+
+    for param_name in params.keys():
+        default_value = params[param_name]
+        for value in test_values:
+            params[param_name] = value
+            try:
+                estimator.set_params(**params)
+            except Exception:
+                # triggered some parameter validation
+                # continue checking other test values
+                pass
+            else:
+                new_params = estimator.get_params()
+                assert_equal(params.keys(), new_params.keys(), msg)
```

Comment:
Not assert_is?

Reply:
Testing for identity doesn't work or keys even if the dictionary really is the same dictionary.
```
>>> d = {'a':1, 'b':2}
>>> d is d
True
>>> d.keys() is d.keys()
False
```

Reply:
Sorry. I misread. But actually, I think that we are not assured that this equality check will work. In py2 this value is a list, in iteration order which is not to be assumed deterministic